### ISSUE TITLE
Documentation relates to Code entropy and the MOD model.

### ISSUE THREAD

**Issue by xekoukou** [sustrik/litmark#1]
@hintjens

Is documentation related to code entropy?

Let me first say that we perceive things because things change. We perceive the notion of color because we see different colors at different times and places. Thus, if there was one color, there would be no word color.

In the same way, we write documentation  on something because the same code can be written differently or because this part will change in the future. If that part of the code never  changed in the future there would be no need for documentation.

Thus, a good programmer needs to identify the parts of the code that have entropy and provide enough information that will make future changes possible.

This is why I believe that model oriented programming is a good methodology to write programs.
We do not only write a program but we also specify a language with which we describe it.

Now, one could say that we should only create a model/template with tags/properties that can take multiple values. We think of the model as a way to customize our application, so that we can create multiple different applications.

In my opinion, the model should contain tags/properties that do not change since we haven't implemented a secondary option but which allow us to represent the inherent entropy of the code itself.

Lets say that our application has a specific architecture that needs to be documented so that a future programmer can be able to change the code.
Even if the architecture never changes, the need for it to be mentioned means that there are other architectures in the wild.

The good with domain languages is that we can create a minimal efficient language that can describe a specific problem. Eskimos have multiple words that represent different types of snow, others do not.

In conclusion, I believe that the MOP can solve the problem you are trying to solve. What we are lacking in https://github.com/sustrik/ribosome is a way to build new domain languages by providing an XML schema and a XML schema validator. 

I wonder whether it would be possible to learn from previous efforts, mostly the XNF tool from imatix.
https://github.com/xekoukou/archive-code/tree/master/Xitami-5.0a0/base2/xnf

Creating a simple xml schema language and a validator seems easy to me.

Using previous schema languages seem very difficult to use, I do not know if they are worth their complexity.
https://en.wikipedia.org/wiki/XML_Schema_(W3C)

**Comment by sustrik** [comment]
What's the problem with reading and validating XML in ribosome?

**Comment by xekoukou** [comment]
ribosome is a code generating tool. Thus it does not provide the tools that would help to write programs in a MOP style.

If we extend our goals to writing good maintainable code in the MOP style, we would need to build a second tool that does a few more things.

A. Create a simple schema language that all new domain languages are written on.
B. Build a schema validator.
C. A program is nothing more than an xml document, so our xml document needs to specify the location of the schema and the dna file that handle it. So we need to build a tool that fetches those files, validates the schema and executes the dna file on this xml document.
D. The xml document can have sections that are described in other xml files that have different xml schemas. So we need to do C for them as well.
E. We need each subschema to specify its dependencies and check the global schema that those dependencies are there.

Apart from the above, I would like to know whether imatix XNF did anything more than that. How did it automate the generation of code generators?

**Comment by sustrik** [comment]
Aren't all those concerns answered by XSD?

As for XNF, I don't really remember. It have been many years since I've seen it.

**Comment by xekoukou** [comment]
Maybe. I looked at the XSD specification and I lost hope that I could use
it(so many pages). Are there any simple tools that do all the previous?

XSD of course specifies a schema language but that is only the first
requirement.

**Comment by sustrik** [comment]
I wonder if an alternative schema language would turn up simpler than XSD.

Anyway, if you are planning to do multiple-level code generation, schemas et c., our exprience back with OpenAMQ was that above certain complexity threshold people cease to be able to follow what's going on.

Thus, try to keep code generation at bay if you want anyone else but you to understand the code.

**Comment by xekoukou** [comment]
Regarding multi-level code generation, my opinion is that the different levels should be done with the use of multiple xml documents and only one level of code generation.

examples:
a) A leg has specific characteristics, and thus we have the schema of the leg.
b) A human body has a different set of characteristics and some of them like the skin color are also a characteristic of the leg.

So we use the xml document of the body to fill some of the characteristics of the leg xml documents.

By doing this, we gain two things:
a)Reduce of complexity of the code generator - only one level.
b) The level of abstraction can be changed at any point, instead of a fixed one. (the .dna file remains the same).


I wonder if you did this as well in openamq or not.
I am also wondering whether you emphasized clarity over productivity. I consider the code generator and MOP as a way to extract the essence of the problem from the lines of code.

Litmark seems to go to the direction of clarity. I suppose that you will feed the generated xml to the code generator. If that is the case, then we have very similar end goals.
